uaclient/api/u/pro/version/v1.py

```diff
@@ -0,0 +1,15 @@
+from uaclient.api.data_types import APIData, APIResponse
+from uaclient.version import get_version
+
+
+def version():
+    response = APIResponse(
+        _schema_version="v1",
+        data=APIData(
+            type="VersionResponse",
+            attributes={"installed_version": get_version()},
+            meta={},
+        ),
+    )
```

Comment:
The module would then look something more like this:
```python
class VersionError(Exception):
    def __init__(self, reason):
        self.reason = reason

class VersionResult(DataObject):
    fields = [
        Field("installed_version", StringDataValue),
    ]

    def __init__(self, *, installed_version: str):
        self.installed_version = installed_version

def version() -> VersionResult:
    try:
        version = get_version()
    except Exception as e:
        raise VersionError(e)
    return VersionResult(installed_version=version)

# this class can be defined in uaclient.api maybe: just an object with a couple attributes
endpoint = APIEndpoint(
    version="1",
    name="Version",
    fn=version,
    options_cls=None,
)
```

And our cli calling code might take
```bash
ua api u.pro.version.v1
```
which would run the following (adapted from the call_api code above)
```python
def call_api(args, *, cfg):
    if args.endpoint_path not in VALID_ENDPOINTS:
        return error_out("invalid endpoint")

    module = import_module("uaclient.api." + args.endpoint_path)
    endpoint = module.endpoint

    try:
        if endpoint.options:
            options = endpoint.options_cls(**args.options)
            result = endpoint.fn(options)
        else:
            result = endpoint.fn()
    except Exception as e:
        return error_out(e)

    response = APIResponse(
        _schema_version=endpoint.version,
        data=APIData(
            type=endpoint.name,
            attributes=result,
            meta={},
        ),
    )
    print(response.to_json())
    return 0
```

Reply:
That makes all sense
I started like this and then ended up moving the higher level object to the endpoint itself - I can see how it makes more sense to have it in the API wrapper.
Thanks for the pointer!

Reply:
Updated basically following @orndorffgrant 's implementation suggestion. Version per se works fine, but maybe the options part still needs to be worked on.
Will add some tests for now based on the current state.

Reply:
@lucasmoura my initial idea about this is that for the API we'd try and separate those actions in steps, and each of those steps can either succeed or fail. Maybe it will not  be possible though, and then we do need to make sure we surface the errors accordingly.